src/main/java/com/palantir/gradle/shadowjar/JarFilesRelocator.java

```diff
@@ -0,0 +1,122 @@
+/*
+ * (c) Copyright 2025 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.gradle.shadowjar;
+
+import com.github.jengelman.gradle.plugins.shadow.relocation.CacheableRelocator;
+import com.github.jengelman.gradle.plugins.shadow.relocation.SimpleRelocator;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Suppliers;
+import com.google.common.collect.ImmutableList;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.Collections;
+import java.util.Set;
+import java.util.function.Supplier;
+import java.util.jar.JarFile;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import java.util.zip.ZipEntry;
+import org.gradle.api.file.FileCollection;
+import org.gradle.api.logging.Logger;
+import org.gradle.api.logging.Logging;
+import org.gradle.api.tasks.Classpath;
+
+/**
+ * Custom relocator that uses {@code Set} based lookups instead of {@link SimpleRelocator}'s include/exclude pattern
+ * matching.
+ *
+ * <p>We don't use {@link SimpleRelocator}'s include list because it uses {@code SelectorUtils.matchPath()} for every
+ * pattern against every file in the jar, which is {@code O(n*m)} where {@code n = number of files} and
+ * {@code m = number of patterns}. This causes a massive performance penalty (e.g., 8 minutes vs 13 seconds for tests).
+ *
+ * <p>Instead, we compute a {@code Set} of relocatable paths upfront and use {@code O(1)} {@code Set} lookups in
+ * {@code canRelocatePath()}.
```

Comment:
```suggestion
 * <p>We don't use {@link SimpleRelocator}'s include list because it uses {@link SelectorUtils#matchPath} for every
 * pattern against every file in the jar, which is {@code O(n*m)} where {@code n = number of files} and
 * {@code m = number of patterns}. This causes a massive performance penalty (e.g., 8 minutes vs 13 seconds for tests).
 *
 * <p>Instead, we compute a set of relocatable paths upfront and use {@code O(1)} set lookups in
 * {@link SimpleRelocator#canRelocatePath}.
```

methods can be referenced in `{@link }` blocks by inserting a hash i.e. `SelectorUtils#matchPath`

As always, read the generated java output in IntelliJ to see what it will look like when people look at the docs!